// @discordjs/rest/dist/lib/CDN.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CDN = void 0;
const constants_1 = require("./utils/constants");
/**
 * The CDN link builder
 */
class CDN {
    constructor(base = constants_1.DefaultRestOptions.cdn) {
        this.base = base;
    }
    /**
     * Generates an app asset URL for a client's asset.
     * @param clientId The client id that has the asset
     * @param assetHash The hash provided by Discord for this asset
     * @param options Optional options for the asset
     */
    appAsset(clientId, assetHash, options) {
        return this.makeURL(`/app-assets/${clientId}/${assetHash}`, options);
    }
    /**
     * Generates an app icon URL for a client's icon.
     * @param clientId The client id that has the icon
     * @param iconHash The hash provided by Discord for this icon
     * @param options Optional options for the icon
     */
    appIcon(clientId, iconHash, options) {
        return this.makeURL(`/app-icons/${clientId}/${iconHash}`, options);
    }
    /**
     * Generates an avatar URL, e.g. for a user or a webhook.
     * @param id The id that has the icon
     * @param avatarHash The hash provided by Discord for this avatar
     * @param options Optional options for the avatar
     */
    avatar(id, avatarHash, { dynamic = false, ...options } = {}) {
        if (dynamic && avatarHash.startsWith('a_')) {
            options.extension = 'gif';
        }
        return this.makeURL(`/avatars/${id}/${avatarHash}`, options);
    }
    /**
     * Generates a banner URL, e.g. for a user or a guild.
     * @param id The id that has the banner splash
     * @param bannerHash The hash provided by Discord for this banner
     * @param options Optional options for the banner
     */
    banner(id, bannerHash, options) {
        return this.makeURL(`/banners/${id}/${bannerHash}`, options);
    }
    /**
     * Generates an icon URL for a channel, e.g. a group DM.
     * @param channelId The channel id that has the icon
     * @param iconHash The hash provided by Discord for this channel
     * @param options Optional options for the icon
     */
    channelIcon(channelId, iconHash, options) {
        return this.makeURL(`/channel-icons/${channelId}/${iconHash}`, options);
    }
    /**
     * Generates the default avatar URL for a discriminator.
     * @param discriminator The discriminator modulo 5
     */
    defaultAvatar(discriminator) {
        return this.makeURL(`/embed/avatars/${discriminator}`);
    }
    /**
     * Generates a discovery splash URL for a guild's discovery splash.
     * @param guildId The guild id that has the discovery splash
     * @param splashHash The hash provided by Discord for this splash
     * @param options Optional options for the splash
     */
    discoverySplash(guildId, splashHash, options) {
        return this.makeURL(`/discovery-splashes/${guildId}/${splashHash}`, options);
    }
    /**
     * Generates an emoji's URL for an emoji.
     * @param emojiId The emoji id
     * @param extension The extension of the emoji
     */
    emoji(emojiId, extension) {
        return this.makeURL(`/emojis/${emojiId}`, { extension });
    }
    /**
     * Generates an icon URL, e.g. for a guild.
     * @param id The id that has the icon splash
     * @param iconHash The hash provided by Discord for this icon
     * @param options Optional options for the icon
     */
    icon(id, iconHash, options) {
        return this.makeURL(`/icons/${id}/${iconHash}`, options);
    }
    /**
     * Generates a guild invite splash URL for a guild's invite splash.
     * @param guildId The guild id that has the invite splash
     * @param splashHash The hash provided by Discord for this splash
     * @param options Optional options for the splash
     */
    splash(guildId, splashHash, options) {
        return this.makeURL(`/splashes/${guildId}/${splashHash}`, options);
    }
    /**
     * Generates a team icon URL for a team's icon.
     * @param teamId The team id that has the icon
     * @param iconHash The hash provided by Discord for this icon
     * @param options Optional options for the icon
     */
    teamIcon(teamId, iconHash, options) {
        return this.makeURL(`/team-icons/${teamId}/${iconHash}`, options);
    }
    /**
     * Constructs the URL for the resource
     * @param base The base cdn route
     * @param options The extension/size options for the link
     */
    makeURL(base, { extension = 'png', size } = {}) {
        extension = String(extension).toLowerCase();
        if (!constants_1.ALLOWED_EXTENSIONS.includes(extension)) {
            throw new RangeError(`Invalid extension provided: ${extension}\nMust be one of: ${constants_1.ALLOWED_EXTENSIONS.join(', ')}`);
        }
        if (size && !constants_1.ALLOWED_SIZES.includes(size)) {
            throw new RangeError(`Invalid size provided: ${size}\nMust be one of: ${constants_1.ALLOWED_SIZES.join(', ')}`);
        }
        const url = new URL(`${this.base}${base}.${extension}`);
        if (size) {
            url.searchParams.set('size', String(size));
        }
        return url.toString();
    }
}
exports.CDN = CDN;
